File: php/107-binary-tree-level-order-traversal2.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution 
{
    private $data = [];
    /**
     * @param TreeNode $root
     * @return Integer[][]
     */
    public function levelOrderBottom($root) 
    {
        if (!$root) {
            return [];
        }
        $queue[] = [$root, 0];
        $data = [];
        while ($queue) {
            $temp = array_shift($queue);
            $node = $temp[0];
            $level = $temp[1];
            $data[$level][] = $node->val;
            if ($node->left) {
                $queue[] = [$node->left, $level+1];
            }
            if ($node->right) {
                $queue[] = [$node->right, $level+1];
            }
        }
        return array_reverse($data);
    }
    //recursive version
    public function levelOrderBottom2($root) 
    {
        $this->traversal($root, 0);           
        return array_reverse($this->data);
    }
    private function traversal($node, $level)
    {
        if (!$node) {
            return;
        }
        $this->data[$level] = $node->val;
        $this->traversal($node->left, $level+1);
        $this->traversal($node->right, $level+1);
    }
}

//test
$node = null;
$s = new Solution();
$s->levelOrderBottom2($node);

File: go/107-binary-tree-level-order-traversal2.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type Node struct {
	tn    *TreeNode
	level int
}

func levelOrderBottom(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var resMap map[int][]int = make(map[int][]int)
	var queue []Node
	queue = append(queue, Node{root, 0})

	for len(queue) > 0 {
		n := queue[0]
		tn := n.tn
		level := n.level
		queue = queue[1:]
		resMap[level] = append(resMap[level], tn.Val)
		if tn.Left != nil {
			queue = append(queue, Node{tn.Left, level + 1})
		}
		if tn.Right != nil {
			queue = append(queue, Node{tn.Right, level + 1})
		}
	}
	var res [][]int

	for i := len(resMap) - 1; i >= 0; i-- {
		res = append(res, resMap[i])
	}

	return res
}
